File: src/frontend/frotas/lib/app/data/model/frota_motorista_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:frotas/app/data/model/model_imports.dart';
import 'package:frotas/app/data/domain/domain_imports.dart';

class FrotaMotoristaModel {
	int? id;
	int? idColaborador;
	String? nome;
	String? numeroCnh;
	String? cnhCategoria;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;

	FrotaMotoristaModel({
		this.id,
		this.idColaborador,
		this.nome,
		this.numeroCnh,
		this.cnhCategoria,
		this.viewPessoaColaboradorModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'nome',
		'numero_cnh',
		'cnh_categoria',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Nome',
		'Numero Cnh',
		'Cnh Categoria',
	];

	FrotaMotoristaModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idColaborador = jsonData['idColaborador'];
		nome = jsonData['nome'];
		numeroCnh = jsonData['numeroCnh'];
		cnhCategoria = FrotaMotoristaDomain.getCnhCategoria(jsonData['cnhCategoria']);
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['nome'] = nome;
		jsonData['numeroCnh'] = numeroCnh;
		jsonData['cnhCategoria'] = FrotaMotoristaDomain.setCnhCategoria(cnhCategoria);
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		nome = plutoRow.cells['nome']?.value;
		numeroCnh = plutoRow.cells['numeroCnh']?.value;
		cnhCategoria = plutoRow.cells['cnhCategoria']?.value != '' ? plutoRow.cells['cnhCategoria']?.value : 'A';
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
	}	

	FrotaMotoristaModel clone() {
		return FrotaMotoristaModel(
			id: id,
			idColaborador: idColaborador,
			nome: nome,
			numeroCnh: numeroCnh,
			cnhCategoria: cnhCategoria,
		);			
	}

	
}

File: backend/frotas/src/model/frota_motorista_model.py
from src import db
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel


class FrotaMotoristaModel(db.Model):
    __tablename__ = 'frota_motorista'

    id = db.Column(db.Integer, primary_key=True)
    nome = db.Column(db.String(100))
    numero_cnh = db.Column(db.String(11))
    cnh_categoria = db.Column(db.String(1))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))

    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_colaborador = data.get('idColaborador')
        self.nome = data.get('nome')
        self.numero_cnh = data.get('numeroCnh')
        self.cnh_categoria = data.get('cnhCategoria')

    def serialize(self):
        return {
            'id': self.id,
            'idColaborador': self.id_colaborador,
            'nome': self.nome,
            'numeroCnh': self.numero_cnh,
            'cnhCategoria': self.cnh_categoria,
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
        }